src/java/org/apache/cassandra/service/accord/api/AccordAgent.java

```diff
@@ -244,73 +258,95 @@ public static long nonClashingStartTime(long startTime, SortedList<Node.Id> node
+    public long expiresAt(ReplyContext replyContext, TimeUnit unit)
     {
-        logger.error(message);
+        return unit.convert(((ResponseContext)replyContext).expiresAtNanos(), NANOSECONDS);
     }
 
-    public TimeoutStrategy slowRead()
+    @Override
+    public long selfSlowAt(TxnId txnId, Status.Phase phase, TimeUnit unit)
     {
-        if (slowRead == null)
+        switch (phase)
         {
-            synchronized (this)
-            {
-                AccordSpec config = DatabaseDescriptor.getAccord();
-                slowRead = new TimeoutStrategy(config.slowRead, TimeoutStrategy.LatencySourceFactory.of(ClientRequestsMetricsHolder.accordReadMetrics));
-            }
+            default: throw new UnhandledEnum(phase);
+            case PreAccept: return unit.convert(slowTxnPreaccept.computeWaitUntil(1), NANOSECONDS);
+            case Execute:   return unit.convert(slowRead.computeWaitUntil(1), NANOSECONDS);
         }
-        return slowRead;
     }
 
-    public TimeoutStrategy slowPreaccept()
+    @Override
+    public long selfExpiresAt(TxnId txnId, Status.Phase phase, TimeUnit unit)
     {
-        if (slowPreaccept == null)
+        long delayNanos;
+        switch (txnId.kind())
         {
-            synchronized (this)
-            {
-                AccordSpec config = DatabaseDescriptor.getAccord();
-                slowPreaccept = new TimeoutStrategy(config.slowPreAccept, TimeoutStrategy.LatencySourceFactory.of(ClientRequestsMetricsHolder.accordReadMetrics));
-            }
+            default: throw new UnhandledEnum(txnId.kind());
+            case Write:
+                delayNanos = DatabaseDescriptor.getWriteRpcTimeout(NANOSECONDS);
+                break;
+            case EphemeralRead:
+            case Read:
+                delayNanos = DatabaseDescriptor.getReadRpcTimeout(NANOSECONDS);
+                break;
+            case ExclusiveSyncPoint:
+                delayNanos = DatabaseDescriptor.getAccordRangeSyncPointTimeoutNanos();
         }
-        return slowPreaccept;
+        return unit.convert(nanoTime() + delayNanos, NANOSECONDS);
+    }
+
+    @Override
+    public AsyncChain<TxnId> awaitStaleId(Node node, TxnId staleId, boolean isRequested)
+    {
+        long waitMicros = (staleId.hlc() + getAccordScheduleDurabilityTxnIdLag(MICROSECONDS)) - node.now();
+        if (waitMicros <= 0)
+            return AsyncChains.success(staleId);
+
+        logger.debug("Waiting {} micros for {} to be stale", waitMicros, staleId);
+        AsyncResult.Settable<TxnId> result = AsyncResults.settable();
+        node.scheduler().selfRecurring(() -> result.setSuccess(staleId), waitMicros, MICROSECONDS);
+        return result;
+    }
+
+    @Override
+    public long minStaleHlc(Node node, boolean requested)
+    {
+        return node.now() - (100 + getAccordScheduleDurabilityTxnIdLag(MICROSECONDS));
```

Comment:
Is 100 chosen here arbitrarily?

Reply:
yes, just some random extra time to mean we don't allocate something always in the future. I guess really it's a bad choice, I didn't think about it much, but if we currently allocate only 1 id every 1000 microseconds this won't actually stop that happening very often.